### ISSUE TITLE
Add an option to load the data in the original order

### ISSUE THREAD

**Issue by pluskid** [tensorflow/datasets#1199]
**Is your feature request related to a problem? Please describe.**

I'm working with collaborators who uses completely different data pipeline that is too much efforts or impossible to switch to tfds. We need to have controlled experiment with aligned example orders in the dataset we loaded. While most of the training procedures only requires i.i.d. samples of the training data, the needs for a well defined order (or an associated `index` property) is also quite common. For example, algorithms such as important sampling or variance reducing for SGD (e.g. SVRG) need to maintain statistics for each example. A well defined order makes it easy to maintain such statistics.

For many dataset, there are well defined ordering from the original raw data (e.g. MNIST or CIFAR). However, tfds generates intermediate data format in tfrecords and the generation process randomly shuffle the original examples. I understand that the shuffling in this step is for some good reason. But this makes collaborating across frameworks very difficult. Moreover, it is not guaranteed that this shuffling ordering will remain the same across different ordering. 

As a result, I'm request a feature to allow loading the data in the original order. Alternatively, loading the data in arbitrary order, but having an `index` property that indicate the order in the original raw data will also work, or maybe even better. Because the latter approach (having an `index` property) makes dealing with largescale dataset that does not fit the whole memory much easier in the context of maintaining statistics for each example.

**Describe the solution you'd like**

An option (or a different function) to load the data set in the original order:

```python
import tensorflow_datasets as tfds
 
ds, info = tfds.load('cifar10', batch_size=-1,
                      as_dataset_kwargs={"shuffle_files": False, "original_order": True},
                      with_info=True)
raw_ds = tfds.as_numpy(ds)
```

Alternatively, the data examples returned by tfds (maybe with an option `with_index=True`) contains an `index` property:

```
batch['image']  # inputs
batch['label']  # labels
batch['index']  # index in the original dataset
```

**Describe alternatives you've considered**

See above.

**Additional context**

**Comment by rsepassi** [comment]
Thanks for this request @pluskid.

There are 2 approaches that come to mind:
1. General: try to do this for all datasets automatically somehow.
2. Specific: add an index feature to datasets for which it makes sense or for which you need it.

General would be tough for a few reasons:
* You'd have to define an ordering semantics that made sense across many file types and datasets with many files, odd directory structures, tables with join keys, etc. This is non-trivial and likely quite brittle.
* If the dataset does not explicitly have an "id" or "idx" feature already, then the DatasetBuilder would have to produce one. A natural one would be to `enumerate` over the records and use the id. It'd be important to ensure that the order in which examples are yielded is deterministic in this case, which is non-trivial to ensure/test.
* Large datasets that are produced with Beam are produced in parallel, and so more work would have to go into ensuring that you produce the correct ids for records.

Specific makes more sense so that things can be special-cased to the datasets for which it makes sense, or users request it.

MNIST, for example, already produces the examples in deterministic order and uses the index as the key for shuffling. It would not be unreasonable to add as a [feature](https://github.com/tensorflow/datasets/blob/00927c8eaa7f2999a2f0a8516b555f28d00f8807/tensorflow_datasets/image/mnist.py#L108) "index" and [yield it in the record](https://github.com/tensorflow/datasets/blob/00927c8eaa7f2999a2f0a8516b555f28d00f8807/tensorflow_datasets/image/mnist.py#L167).

For datasets that do this, I think we should probably require that the index also be used as the shuffle key.

If users request and implement this for many datasets, then we could think about a more general approach.

**Comment by pluskid** [comment]
@rsepassi Thanks for the reply. I totally agree with you that it is difficult to define an index (that everybody will be happy with) for largescale datasets that do not naturally come with one, and especially when the datasets are built with distributed processing pipelines.

I like your suggestion that for now, it is good to only add support (or have an option to add support) for datasets that already have a naturally defined order. In particular, I think having a feature `index` and yield it with other features will be the most flexible that will go well with shuffling.

Are you suggesting that this optional `index` feature will be considered to be added to selected datasets? Or do you mean, say, if I need MNIST with pre-defined ordering, I should make a copy of the existing MNIST dataset code and hack it to add this `index` feature, and so on for other datasets?

**Comment by Conchylicultor** [comment]
@pluskid, you can send a PR to modify the existing mnist dataset to add this `index` key.  I believe that is what rsepassi is suggesting.
But the user would still have to manually sort the dataset in memory using this "index" key. This would works for small dataset like Cifar/Mnist, but won't be scalable to bigger datasets.

**Comment by pluskid** [comment]
@rsepassi @Conchylicultor Thanks for the clarification and suggestions on how to add this feature. I'm closing this issue for now as it has served its purpose. I will try to send a PR if I find a chance to do that in the future.

**Comment by pluskid** [comment]
Today I found that the following code snippet produce different results when connected to two different colab kernel (one several days older than the other)

```
data_name = 'mnist' 
raw_data, info = tfds.load(name=data_name, batch_size=-1, with_info=True,
                           as_dataset_kwargs={'shuffle_files': False})
raw_data = tfds.as_numpy(raw_data)
print(raw_data['train']['label'][:10])
```

What could be the potential reason? Is that a new version of the dataset get generated in tfds and the order changed again? Or some other potential issue?

The two different results are:

```
[4 1 0 7 8 1 2 7 1 6]
```

and 

```
[8 8 3 0 0 4 0 6 1 1]
```

**Comment by Conchylicultor** [comment]
Thanks for reporting. Could you make sure you have the last version of TFDS installed ? Colab may use an older version but you can update it within colab with `!pip`.
Also could you try to explicitly pin the version to `data_name = 'mnist:3.0.0'` ?

**Comment by pluskid** [comment]
@Conchylicultor Thanks for the tip! The problem is that a subset of my results are analyzed using an old colab (i.e. analyzed a few days/weeks ago). So if I update I will have to re-run all the old analysis. And I'm using internal colab so pip is probably not available.

I will try to pin the version. Is it expected that if the dataset version change, the order of the example will be different? Thanks!

**Comment by Conchylicultor** [comment]
We updated the generation such as each example is uniquely identified by a key. The key is used to shuffle examples deterministically. Re-generating the dataset or updating the dataset shouldn't modify the order (as long as the number of example stays the same and keys are the same).
Old versions of TFDS (previous to tfds 2.0.0) do not have this system (at least not by default), so may have different generation order.

**Comment by pierrot0** [comment]
A few months ago, we also added the "disable_shuffling" option (see https://www.tensorflow.org/datasets/add_dataset).
